SignalServiceKit/src/Storage/OWSStorage.m

```diff
@@ -539,36 +656,61 @@ - (NSData *)databasePassword
             }
         } else {
             [self backgroundedAppDatabasePasswordInaccessibleWithErrorDescription:
-                      @"Password inaccessible; not main app."];
+                      [NSString stringWithFormat:@"%@ inaccessible; not main app.", label]];
         }
 
         // At this point, either this is a new install so there's no existing password to retrieve
         // or the keychain has become corrupt.  Either way, we want to get back to a
         // "known good state" and behave like a new install.
 
-        BOOL shouldHavePassword = [NSFileManager.defaultManager fileExistsAtPath:[self databaseFilePath]];
-        if (shouldHavePassword) {
+        BOOL shouldHaveDatabaseMetadata = [NSFileManager.defaultManager fileExistsAtPath:[self databaseFilePath]];
+        if (shouldHaveDatabaseMetadata) {
             OWSProdCritical([OWSAnalyticsEvents storageErrorCouldNotLoadDatabaseSecondAttempt]);
         }
 
         // Try to reset app by deleting database.
         [OWSStorage resetAllStorage];
 
-        dbPassword = [self createAndSetNewDatabasePassword];
+        data = createDataBlock();
     }
 
-    return [dbPassword dataUsingEncoding:NSUTF8StringEncoding];
+    return data;
 }
 
-- (NSString *)createAndSetNewDatabasePassword
+- (NSData *)createAndSetNewDatabasePassword
 {
-    NSString *password = [[Randomness generateRandomBytes:30] base64EncodedString];
+    NSData *password = [[[Randomness generateRandomBytes:kDatabasePasswordLength] base64EncodedString]
+        dataUsingEncoding:NSUTF8StringEncoding];
 
     [OWSStorage storeDatabasePassword:password];
 
     return password;
 }
 
+- (NSData *)createAndSetNewDatabaseSalt
+{
+    NSData *saltData = [Randomness generateRandomBytes:(int)kSQLCipherSaltLength];
+
+    [OWSStorage storeDatabaseSalt:saltData];
+
+    return saltData;
+}
+
+- (NSData *)createAndSetNewDatabaseKeySpec
+{
+    NSData *databasePassword = [self databasePassword];
+    OWSAssert(databasePassword.length > 0);
+    NSData *databaseSalt = [self databaseSalt];
+    OWSAssert(databaseSalt.length == kSQLCipherSaltLength);
+
+    NSData *keySpecData = [YapDatabaseCryptoUtils databaseKeySpecForPassword:databasePassword saltData:databaseSalt];
+    OWSAssert(keySpecData.length == kSQLCipherKeySpecLength);
```

Comment:
We never use a random key spec - it's always derived from the password + salt so that all three are always valid and interoperable.